src/sage/groups/generic.py

```diff
@@ -458,6 +459,7 @@ def bsgs(a, b, bounds, operation='*', identity=None, inverse=None, op=None):
         inverse = inv
         op = mul
     elif operation in addition_names:
+        # Should this be replaced with .zero()? With an extra AttributeError handler?
         identity = a.parent()(0)
```

Comment:
I think `.zero()` should be faster, but indeed not all additive groups have a `.zero()` method...

Reply:
I took the violent approach of replacing it with `.zero()` and fixing the errors. I only found a few use cases of `sage.groups.generic` in the entire Sage, so I fixed those (only required fixing `homset::zero`)

Reply:
Thanks, looks promising! Just double-checking: The change from `.codomain()` to `.extended_codomain()` in `src/sage/schemes/generic/homset.py` is an essentially unrelated bugfix, right? And this is what causes the point `(a : a)` in that doctest in `src/sage/schemes/projective/projective_morphism.py` to be normalized properly?